src/main/java/com/snowflake/kafka/connector/internal/streaming/schemaevolution/iceberg/IcebergSchemaEvolutionService.java

```diff
@@ -17,48 +20,163 @@ public class IcebergSchemaEvolutionService implements SchemaEvolutionService {
+    return columnNameWithNestedNames.split("\\.")[0];
+  }
+
+  private String removeDoubleQuotes(String columnName) {
+    return columnName.replaceAll("\"", "");
+  }
+
+  /** Columns that are not present in a current schema are to be added */
+  private List<IcebergColumnTree> distinguishColumnsToAdd(
+      List<IcebergColumnTree> alreadyExistingColumns,
+      List<IcebergColumnTree> modifiedOrAddedColumns) {
+    return modifiedOrAddedColumns.stream()
+        .filter(
+            modifiedOrAddedColumn ->
+                alreadyExistingColumns.stream()
+                    .noneMatch(
+                        tree ->
+                            tree.getColumnName()
+                                .equalsIgnoreCase(modifiedOrAddedColumn.getColumnName())))
+        .collect(Collectors.toList());
+  }
+
+  /** If columns is present in a current schema it means it has to be modified */
+  private List<IcebergColumnTree> distinguishColumnsToModify(
+      List<IcebergColumnTree> alreadyExistingColumns,
+      List<IcebergColumnTree> modifiedOrAddedColumns) {
+    return modifiedOrAddedColumns.stream()
+        .filter(
+            modifiedOrAddedColumn ->
+                alreadyExistingColumns.stream()
+                    .anyMatch(
+                        tree ->
+                            tree.getColumnName()
+                                .equalsIgnoreCase(modifiedOrAddedColumn.getColumnName())))
+        .collect(Collectors.toList());
+  }
+
+  private void alterAddColumns(String tableName, List<IcebergColumnTree> addedColumns) {
+    if (addedColumns.isEmpty()) {
+      return;
     }
+    Map<String, ColumnInfos> columnInfosMap = toColumnInfos(addedColumns);
+    try {
+      conn.appendColumnsToIcebergTable(tableName, columnInfosMap);
+    } catch (SnowflakeKafkaConnectorException e) {
+      logQueryFailure(tableName, e);
+    }
+  }
+
+  private void alterDataType(
+      String tableName,
+      List<IcebergColumnTree> alreadyExistingColumns,
+      List<IcebergColumnTree> modifiedColumns) {
+    if (modifiedColumns.isEmpty()) {
+      return;
+    }
+    mergeChangesIntoExistingColumns(alreadyExistingColumns, modifiedColumns);
+    Map<String, ColumnInfos> columnInfosMap = toColumnInfos(alreadyExistingColumns);
+    try {
+      conn.alterColumnsDataTypeIcebergTable(tableName, columnInfosMap);
+    } catch (SnowflakeKafkaConnectorException e) {
+      logQueryFailure(tableName, e);
+    }
+  }
+
+  private void mergeChangesIntoExistingColumns(
+      List<IcebergColumnTree> alreadyExistingColumns, List<IcebergColumnTree> modifiedColumns) {
+    alreadyExistingColumns.forEach(
+        existingColumn -> {
+          IcebergColumnTree mewVersion =
+              modifiedColumns.stream()
+                  .filter(c -> c.getColumnName().equals(existingColumn.getColumnName()))
+                  .collect(Collectors.toList())
+                  .get(0);
```

Comment:
I am not so sure that we don't end up with `ArrayIndexOutOfBoundsException` here. Please add some meaningful logging for that case.

Reply:
Yea I agree. It's too optimistic I'll change it.